### PR TITLE
Added support for custom ssl client cert in grpc_cli

### PR DESCRIPTION
**Pull request by farcaller**
This extends grpc_cli with support for client cert auth (a very naive, but functional implementation).

### CODE REVIEW

**Comment by nicolasnoble** [comment]
@grpc-jenkins: test this please.

File: test/cpp/util/cli_credentials.cc [hunk 1]
```diff
@@ -49,7 +54,31 @@ std::shared_ptr<grpc::ChannelCredentials> CliCredentials::GetCredentials()
     if (FLAGS_use_auth) {
       return grpc::GoogleDefaultCredentials();
     } else {
-      return grpc::SslCredentials(grpc::SslCredentialsOptions());
+      auto options = grpc::SslCredentialsOptions();
+      if (!FLAGS_ssl_server_ca_file.empty()) {
+        std::stringstream ca_buffer;
+        ca_buffer << std::ifstream(FLAGS_ssl_server_ca_file).rdbuf();
+
```
Comment (jboeuf) [inline test/cpp/util/cli_credentials.cc]: please remove blank line here.

File: test/cpp/util/cli_credentials.cc [hunk 2]
```diff
@@ -33,10 +33,15 @@
 
 #include "test/cpp/util/cli_credentials.h"
 
+#include <fstream>
+#include <sstream>
 #include <gflags/gflags.h>
 
 DEFINE_bool(enable_ssl, false, "Whether to use ssl/tls.");
 DEFINE_bool(use_auth, false, "Whether to create default google credentials.");
+DEFINE_string(ssl_server_ca_file, "", "Path to server CA file.");
```
Comment (jboeuf) [inline test/cpp/util/cli_credentials.cc]: `ssl_server_roots_file, "", "Path to custom server roots file. If this option is not specified, the default server roots will be used")`

File: test/cpp/util/cli_credentials.cc [hunk 3]
```diff
@@ -49,7 +54,31 @@ std::shared_ptr<grpc::ChannelCredentials> CliCredentials::GetCredentials()
...
+      if (!FLAGS_ssl_server_ca_file.empty()) {
+        std::stringstream ca_buffer;
+        ca_buffer << std::ifstream(FLAGS_ssl_server_ca_file).rdbuf();
+
+        if (ca_buffer.str().empty()) {
+          exit(2);
+        }
+        options.pem_root_certs = ca_buffer.str();
+      }
+      if (!FLAGS_ssl_client_cert_file.empty() && !FLAGS_ssl_client_key_file.empty()) {
```
Comment (jboeuf) [inline test/cpp/util/cli_credentials.cc]: Please handle the case when was is empty and the other is not and warn the user that this is an invalid combination.

File: test/cpp/util/cli_credentials.cc [hunk 4]
```diff
@@ -33,10 +33,15 @@
...
 #include "test/cpp/util/cli_credentials.h"
 
+#include <fstream>
+#include <sstream>
 #include <gflags/gflags.h>
 
 DEFINE_bool(enable_ssl, false, "Whether to use ssl/tls.");
 DEFINE_bool(use_auth, false, "Whether to create default google credentials.");
+DEFINE_string(ssl_server_ca_file, "", "Path to server CA file.");
+DEFINE_string(ssl_client_cert_file, "", "Path to client certificate file.");
```
Comment (jboeuf) [inline test/cpp/util/cli_credentials.cc]: `ssl_client_cert_chain_file` (and associated comment to specify that this is a chain).

File: test/cpp/util/cli_credentials.cc (hunk 2 above)
Comment (jboeuf) [inline test/cpp/util/cli_credentials.cc]: For these flags, please specify that --enable_ssl must be used and check that this is the case in the code.

Comment (farcaller) [inline test/cpp/util/cli_credentials.cc]: Done.

Comment (farcaller) [inline test/cpp/util/cli_credentials.cc]: Done.

File: test/cpp/util/cli_credentials.cc (hunk 4 above)
Comment (farcaller) [inline test/cpp/util/cli_credentials.cc]: Done.

File: test/cpp/util/cli_credentials.cc (hunk 1 above)
Comment (farcaller) [inline test/cpp/util/cli_credentials.cc]: Done.

File: test/cpp/util/cli_credentials.cc (hunk 3 above)
Comment (farcaller) [inline test/cpp/util/cli_credentials.cc]: Done. Also added error message below.

File: test/cpp/util/cli_credentials.cc [hunk 5]
```diff
@@ -49,7 +56,40 @@ std::shared_ptr<grpc::ChannelCredentials> CliCredentials::GetCredentials()
     if (FLAGS_use_auth) {
```
Comment (jboeuf) [inline test/cpp/util/cli_credentials.cc]: Some flags combinations are still meaningless and error prone. I would have a helper function like:

```cpp
void CheckForInvalidSslFlags(const string& flag) {
  if (!FLAGS_ssl_cleint_cert_chain_file.empty() || !FLAGS_...) {
    std::cerr << "SSL specific flags cannot be used with " << flag << " flag." << std::endl;
    exit(2);
  }
}
```
and call this from:

```cpp
if (!FLAGS_enable_ssl) {
  CheckForInvalidSslFlags("--noenable_ssl");
  return grpc::InsecureCredentials();
} else {
  if (FLAGS_use_auth) {
    CheckForInvalidSslFlags("--use_auth");
    return grpc::GoogleDefaultCredentials();
  } 
...
```

Comment (farcaller) [inline test/cpp/util/cli_credentials.cc]: oops, it dropped off my radar. I'll fix that tomorrow.

**Comment by farcaller** [comment]
PTAL?

**Comment by jtattermusch** [comment]
test this please

**Comment by grpc-testing** [comment]
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

**Comment by grpc-testing** [comment]
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

**Comment by pxLi** [comment]
Any update on this one? Thanks~

**Comment by farcaller** [comment]
Oh hm. I need to find where did I stash the sources and actually fix the typo.

**Comment by farcaller** [comment]
Fixed the typo, PTAL?

**Comment by mhaidrygoog** [comment]
Thanks for the PR! Your changes are approved. Please resolve conflicts and we'll merge on a clean test run

**Comment by farcaller** [comment]
@mhaidrygoog how's that approved if the similar functionality [was merged in many months ago](https://github.com/grpc/grpc/commit/c173b196d66430d96cd52402707bd874fc2931db), making this PR obsolete? Am I confusing something?

**Pull request closed without merging**